devolutions-gateway/src/streaming/mod.rs

```diff
@@ -0,0 +1,69 @@
+use anyhow::Context;
+use axum::{body::Body, response::Response};
+use streamer::{webm_stream, ReOpenableFile};
+use uuid::Uuid;
+
+use crate::{recording::OnGoingRecordingState, token::RecordingFileType, ws::websocket_compat};
+
+struct ShutdownSignal(devolutions_gateway_task::ShutdownSignal);
+
+impl streamer::Signal for ShutdownSignal {
+    fn wait(&mut self) -> impl std::future::Future<Output = ()> + Send {
+        self.0.wait()
+    }
+}
+
+pub(crate) async fn stream_file(
+    path: camino::Utf8PathBuf,
+    ws: axum::extract::WebSocketUpgrade,
+    shutdown_signal: devolutions_gateway_task::ShutdownSignal,
+    recordings: crate::recording::RecordingMessageSender,
+    recording_id: Uuid,
+) -> anyhow::Result<Response<Body>> {
+    // 1.identify the file type
+    if path.extension() != Some(RecordingFileType::WebM.extension()) {
+        return Err(anyhow::anyhow!("Invalid file type"));
+    }
+    // 2.if the file is actively being recorded, then proceed
+    let Ok(Some(OnGoingRecordingState::Connected)) = recordings.get_state(recording_id).await else {
+        return Err(anyhow::anyhow!("File is not being recorded"));
+    };
+
+    let streaming_file = ReOpenableFile::open(&path).with_context(|| format!("failed to open file: {path:?}"))?;
+
+    let upgrade_result = ws.on_upgrade(move |socket| async move {
+        let websocket_stream = websocket_compat(socket);
+        // Spawn blocking because webm_stream is blocking
+        let streaming_result = tokio::task::spawn_blocking(move || {
+            webm_stream(
+                websocket_stream,
+                streaming_file,
+                ShutdownSignal(shutdown_signal),
+                // Question: Do we need to make it configurable
+                streamer::StreamingConfig { encoder_threads: 4 },
+                move || {
+                    let (tx, rx) = tokio::sync::oneshot::channel();
+                    recordings
+                        .add_new_chunk_listener(recording_id, tx)
+                        .expect("Failed to send on_appended message"); // early development
```

Comment:
Can you double check this comment? Should expect be removed?

Reply:
updated